spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java

```diff
@@ -220,6 +258,9 @@ private void addChannelFromString(Collection<MessageChannel> channels, String ch
 		MessageChannel channel = resolveChannelForName(channelName, message);
 		if (channel != null) {
 			channels.add(channel);
+			if (!mapped && !(this.dynamicChannels.get(channelName) != null)) {
+				this.dynamicChannels.put(channelName, channel);
```

Comment:
Please read the javadocs for LinkedHashMap when using that ctor - `get()` updates the LRU 😄 .